cdhweb/projects/migrations/0011_calculate_membership_dates.py

```diff
@@ -0,0 +1,108 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.23 on 2020-11-10 17:10
+from __future__ import unicode_literals
+
+from django.db import migrations
+from django.core.exceptions import ObjectDoesNotExist
+
+
+def calculate_membership_dates(apps, schema_editor):
+    # We can't import the Person model directly as it may be a newer
+    # version than this migration expects. We use the historical version.
+    Membership = apps.get_model('projects', 'Membership')
+    # sort by project, then user, then grant start date (role?)
+    memberships = Membership.objects.all() \
+                            .order_by('project', 'user', 'grant__start_date')
+
+    # as we loop through, track current membership to extend across grants
+    current = None
+    for membership in memberships:
+        if current:
+            # if this membership matches the previous one, extend it
+            if current.user == membership.user and \
+               current.project == membership.project and \
+               current.role == membership.role:
+                # extend the end date
+                current.end_date = calculate_membership_end(membership)
+                # delete the obsolete membership
+                membership.delete()
+                continue
+
+            # if it doesn't match, save it
+            else:
+                current.save()
+                current = None
+
+        if not current:
+            # if there is no current membership to extend, start a new one
+            current = membership
+            # set dates based on associated grant & cdh position dates
+            current.start_date = calculate_membership_start(membership)
+            current.end_date = calculate_membership_end(membership)
+
+    # save the last record after loop ends
+    if current:
+        current.save()
+
+    # if any memberships are missing start dates, it will be a problem
+    assert not Membership.objects.filter(start_date__isnull=True).exists()
```

Comment:
should we just do this at the top of the function? might as well bail if there's nothing to do

Reply:
Since it's a migration, it should only run when we need it, right after start and end date fields have been created and are null.